vm/src/vm/runners/cairo_pie.rs

```diff
@@ -68,6 +79,124 @@ pub struct CairoPie {
+const N_OFFSET_BITS: usize = 47;
+const SEGMENT_MASK: u64 = ((1 << N_SEGMENT_BITS) - 1) << N_OFFSET_BITS;
+const OFFSET_MASK: u64 = (1 << N_OFFSET_BITS) - 1;
+
+fn maybe_relocatable_from_le_bytes(bytes: &[u8]) -> MaybeRelocatable {
+    // Little-endian -> the relocatable bit is in the last element
+    let is_relocatable = (bytes[bytes.len() - 1] & 0x80) != 0;
+
+    if !is_relocatable {
+        let felt = Felt252::from_bytes_le(bytes);
+        return MaybeRelocatable::Int(felt);
+    }
+
+    // Relocatable values are guaranteed to fit in a u64
+    let value = {
+        let mut value = 0;
+        for (index, byte) in bytes[..8].iter().enumerate() {
+            value += u64::from(*byte) << (index * 8);
+        }
+        value
+    };
+
+    let segment = (value & SEGMENT_MASK) >> N_OFFSET_BITS;
+    let offset = value & OFFSET_MASK;
+    MaybeRelocatable::RelocatableValue(Relocatable::from((segment as isize, offset as usize)))
+}
+
+fn read_memory_file<R: Read>(
+    mut reader: R,
+    addr_size: usize,
+    felt_size: usize,
+) -> Result<CairoPieMemory, DeserializeMemoryError> {
+    let memory_cell_size = addr_size + felt_size;
+    let mut memory = CairoPieMemory::new();
+    let mut pos: usize = 0;
+
+    loop {
+        let mut element = vec![0; memory_cell_size];
+        match reader.read(&mut element) {
+            Ok(n) => {
+                if n == 0 {
+                    break;
+                }
+                if n != memory_cell_size {
+                    return Err(DeserializeMemoryError::UnexpectedEof);
+                }
+            }
+            Err(e) => return Err(e.into()),
+        }
+        let (address_bytes, value_bytes) = element.split_at(addr_size);
+        let address = maybe_relocatable_from_le_bytes(address_bytes.clone());
+        let value = maybe_relocatable_from_le_bytes(value_bytes);
+
+        match address {
+            MaybeRelocatable::RelocatableValue(relocatable) => {
+                memory.push((
+                    (relocatable.segment_index as usize, relocatable.offset),
+                    value,
+                ));
+            }
+            MaybeRelocatable::Int(_value) => {
+                return Err(DeserializeMemoryError::AddressIsNotRelocatable(pos));
+            }
+        }
+        pos += memory_cell_size;
+    }
+
+    Ok(memory)
+}
+
+impl CairoPie {
+    #[cfg(feature = "std")]
+
+    pub fn from_zip_archive<R: Read + Seek>(
+        mut zip: zip::ZipArchive<R>,
+    ) -> Result<CairoPie, CairoPieError> {
+        let metadata: CairoPieMetadata = parse_zip_file(zip.by_name("metadata.json")?)?;
+        let execution_resources: ExecutionResources =
+            parse_zip_file(zip.by_name("execution_resources.json")?)?;
+        let additional_data: HashMap<String, BuiltinAdditionalData> =
+            parse_zip_file(zip.by_name("additional_data.json")?)?;
+        let version: CairoPieVersion = parse_zip_file(zip.by_name("version.json")?)?;
```

Comment:
It looks like all of these `parse_zip_file()` calls will redundantly parse the zip file, is there an easy way to avoid this?